src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs

```diff
@@ -0,0 +1,361 @@
+
+            var elements = new List<ProjectItemElement>();
+
+            foreach (var item in project.AllEvaluatedItems)
+            {
+                if (lookup.Contains(unconfiguredProject.MakeRooted(item.EvaluatedInclude)))
+                {
+                    elements.Add(item.Xml);
+                }
+            }
+
+            return elements;
+        }
+
+        /// <summary>
+        /// Gets the display order for a project tree.
+        /// </summary>
+        private static int GetDisplayOrder(IProjectTree projectTree)
+        {
+            if (projectTree is IProjectTree2 projectTree2)
+            {
+                return projectTree2.DisplayOrder;
+            }
+            // It's safe to return zero here. Project trees that do not have a display order are always assumed zero.
+            return 0;
+        }
+
+        /// <summary>
+        /// Checks to see if the display order is valid.
+        /// </summary>
+        private static bool IsValidDisplayOrder(int displayOrder)
+        {
+            return displayOrder > 0 && displayOrder != int.MaxValue;
+        }
+
+        /// <summary>
+        /// Gets a sibling based on the given project tree. Can return null.
+        /// </summary>
+        /// <param name="projectTree">the given project tree</param>
+        /// <param name="returnSibling">passes the index of the given project tree from the given ordered sequence, expecting to return a sibling</param>
+        /// <returns>a sibling</returns>
+        private static IProjectTree2 GetSiblingByDisplayOrder(IProjectTree projectTree, Func<int, IEnumerable<IProjectTree>, IProjectTree2> returnSibling)
+        {
+            var parent = projectTree.Parent;
+            var displayOrder = GetDisplayOrder(projectTree);
+            if (!IsValidDisplayOrder(displayOrder) || parent == null)
+            {
+                return null;
+            }
+
+            var orderedChildren = parent.Children.Where(x => IsValidDisplayOrder(GetDisplayOrder(x))).OrderBy(x => GetDisplayOrder(x)).ToArray();
+
+            for (var i = 0; i < orderedChildren.Length; ++i)
+            {
+                var sibling = orderedChildren[i];
+                if (GetDisplayOrder(sibling) == displayOrder)
+                {
+                    return returnSibling(i, orderedChildren);
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the previous sibling of the given project tree, if there is any. Can return null.
+        /// </summary>
+        private static IProjectTree2 GetPreviousSibling(IProjectTree projectTree)
+        {
+            return GetSiblingByDisplayOrder(projectTree, (i, orderedChildren) =>
+            {
+                if (i == 0)
+                {
+                    return null;
+                }
+
+                return orderedChildren.ElementAt(i - 1) as IProjectTree2;
```

Comment:
Does this `ElementAt` end up using an indexer, or does it cause *another* walk through the tree?

Reply:
It will use an index. https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/ElementAt.cs#L28